readme.md
# yt-dl-service
A minimalistic wrapper for `yt-dlp` which is designed to run as a service and keep track of e.g. youtube channels.
If there's a risk that certain channel can be taken down, this utility helps to keep up-to-date backup locally.

This is achieved by running `yt-dlp` in background for a list of links according to schedule.

> *DISCLAIMER: This is a personal project, tailored to my needs. Therefore
> no PKGBUILD or prebuilt binaries are provided just yet. Follow
> [manual installation](#manual-installation) if you want to continue.*

## Manual Installation
1. **Get [yt-dlp](https://github.com/yt-dlp/yt-dlp)**, depending on your platform. For example on Arch:
```sh
$ sudo pacman -S yt-dlp
```
> As a result of installation, `yt-dlp` must be available in PATH - check by running `yt-dlp --version`.

2. **Get rust toolchain** if you don't have it already, usually this is done via [rustup](https://rustup.rs/). For example on Arch:
```sh
$ sudo pacman -S rustup
$ rustup toolchain install stable
```

3. **Build executable**
```sh
$ cd yt-dl-service
$ cargo install --path . --root ~/.cargo
```
> Binary `yt-dl-service` will be located in `~/.cargo/bin/`

4. **Configure service**

4.1 Copy necessary files
```sh
$ sudo mkdir /etc/yt-dl-service
$ sudo mkdir /var/log/yt-dl-service 
$ sudo cp ./templates/settings.json /etc/yt-dl-service/
$ sudo cp ./templates/yt-dl-service.service /etc/systemd/system/
$ sudo cp ./templates/yt-dl-service.timer /etc/systemd/system/
```
4.2 Change username in service
```sh
$ sudo <YOUR FAVORITE EDITOR> /etc/systemd/system/yt-dl-service.service
```

4.3 Customize settings, refer to [settings chapter](#settings). At this point you want add desired time (UTC tz) in `updateSchedule` and add some channels to `tasks`.
```sh
$ sudo <YOUR FAVORITE EDITOR> /etc/yt-dl-service/settings.json
```
Also be sure to set `"updateOnStart": true` on first launch in order to perform initial download and check if everything is working ok. Don't forget to set it back to `false` once you've ensured it's working.

> Alternatively, you can keep it always on. In that case channels will be updated on boot and when you restart the service.

> WARNING: output files path shouldn't be bound to any user, as the service can 
> start before user login. E.g. if output path is (an external) drive, which is
> mounted after login, service will attempt to write into inexistant drive, and
> after login actual drive would be mounted in different path. To solve this you
> can either add your drive in fstab to have permanent mount location or run the 
> service with --user flag (additional changes to service file and its location are
> needed).

4.4 Enable systemd service
```sh
$ sudo systemctl enable --now yt-dl-service.timer
``` 
> Additional timer service allows to offset start time of actual service by 1 min,
> this is likely unnecessary but other system components can properly load.  

5. **Summary**

As a result, you should have:
* binary `/home/username/.cargo/bin/yt-dl-service`
* config `/etc/yt-dl-service/settings.json`
* log `/var/log/yt-dl-service/yt-dl-service.log`
* systemd service `/etc/systemd/system/yt-dl-service.service`
* systemd timer `/etc/systemd/system/yt-dl-service.timer`

6. **Adding new channels**

Modify settings file and reload
```sh
sudo systemctl restart yt-dl-service.service
```

## Settings
Configuration is done solely via `settings.json` file.
Generally, it should look like this:
```json
{
    "tasks": [
        {
            "url": "<YT CHANNEL URL A>",
            "outputPath": "/home/myusername/channels-backup/channel-a",
            "name": "This is A"
        },
        {
            "url": "<YT CHANNEL URL B>",
            "outputPath": "/home/myusername/channels-backup/channel-b",
            "name": "This is B"
        },
    ],
    "updateSchedule": [{"daily":"20:30"}],
    "updateOnStart": true,
    "setOwner": "myusername",
    "setGroup": "myusername" 
}
```

- `tasks` - list of tasks to be executed every update time
- `updateSchedule` - schedule when updates happen
- `updateOnStart` - if *true*, regardless of schedule the update will happen on program startup, e.g. on boot
- `setOwner` and `setGroup` - set owner and group for downloaded files, **NOTE**: this applies to all files in output dirs, not only new or produced by yt-dl-service

### Example, make schedule:
**Note that all time is in UTC, not your local time.**
Global setting, all tasks are executed at specified time(s).
Can be either `daily` with single time or `cron` with custom time.
In example below, it will execute daily at 12:00 and 18:30 UTC, also at 8:00 on mondays of year 2049.
If you are not familiar with `cron`, refer [here](https://en.wikipedia.org/wiki/Cron) and directly to [crate](https://crates.io/crates/tokio-cron-scheduler) docs.
```json
{
    "tasks": [
        ...
    ],
    "updateSchedule": [
        { "daily": "12:00" },
        { "daily": "18:30" },
        { "cron": "0 0 8 * * 1 2049" }
    ],
    ...
}
```

### Example, download only audio:
Task-specific setting, if `audioFormat` field is not present, will default to mp3.
Possible values for `audioFormat` are: mp3, aac, m4a, opus, vorbis, flac, alac, wav
```json
"tasks": [
    {
        ...
        "audioOnly": true,
        "audioFormat": "wav"
    },
    ...
],
```

### Example, limit video resolution:
Task-specific setting.
Examples for `maxResolution`: 480, 720, 1080, 1440, 2160, 4320 ...
```json
"tasks": [
    {
        ...
        "maxResolution": 720
    },
    ...
],
```


### Example, limit download speed:
Task-specific setting.
Examples for `throttleSpeed`: 3.5M, 300K ...
```json
"tasks": [
    {
        ...
        "throttleSpeed": "500K"
    },
    ...
],
```

### Example, custom yt-dlp flags:
Task-specific setting.
Pass any flags supported by yt-dlp to achieve desired behavior.
```json
"tasks": [
    {
        ...
        "customFlags": ["-f", "bestvideo[height=480]+bestaudio"]
    },
    ...
],
```


CHANGELOG
# Changelog

## [Unreleased]

## [0.2.0] - 18-08-2022
- Set file owner after download
- Remove tmp dir

## [0.1.0] - 16-07-2022
- Initial MVP version: downloads, basic config and scheduler

src/main.rs
use settings::{validate_settings, LogLevel};
use std::env;
use tracing::metadata::LevelFilter;
use tracing::{debug, error, info};
use tracing_subscriber::prelude::*;
use watcher::Watcher;

mod settings;
mod watcher;

static DEFAULT_SETTINGS_PATH: &str = "./settings.json";
static DEFAULT_LOG_DIR: &str = "./";
static LOG_NAME: &str = "yt-dl-service.log";

fn setup_logger(log_dir: &str, log_name: &str, level: LogLevel) -> impl Drop {
    let (file_nb, guard) =
        tracing_appender::non_blocking(tracing_appender::rolling::never(log_dir, log_name));
    let level: LevelFilter = level.into();
    let fmt_layer = tracing_subscriber::fmt::layer()
        .with_ansi(false)
        .with_writer(file_nb)
        .with_filter(level);
    tracing_subscriber::registry().with(fmt_layer).init();
    guard
}

#[tokio::main]
async fn main() {
    // Get cli args
    let args: Vec<String> = env::args().collect();
    let settings_path = args
        .get(1)
        .map(String::as_str)
        .unwrap_or(DEFAULT_SETTINGS_PATH);
    // Load settings from file
    let s = settings::load_settings(settings_path);
    let log_path = s.log_dir.as_deref().unwrap_or(DEFAULT_LOG_DIR);
    let log_level = s.log_level.clone().unwrap_or(LogLevel::Info);
    let _guard = setup_logger(log_path, LOG_NAME, log_level);
    debug!("{:?}", s);
    if let Err(e) = validate_settings(&s) {
        error!("Stopping, failed to validate settings: {}", e);
        return;
    }
    // Main loop
    info!("Starting watcher.");
    if let Err(e) = Watcher::new(s).run().await {
        error!("Stopping, runtime error: {}", e);
    }
}


src/watcher.rs
use crate::settings::{parse_time, DownloadEntity, ScheduledTime, Settings};
use std::cell::Cell;
use std::collections::HashSet;
use std::fs::{read_dir, remove_dir};
use std::path::{Path, PathBuf};
use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use tokio::sync::{mpsc, oneshot};
use tokio_cron_scheduler::{Job, JobScheduler};
use tracing::{debug, error, info, trace, warn};

type VideoId = String;

/// Temp files are downloaded to tmp directory located inside given output dir.
static TEMP_DIR: &str = "tmp";

pub fn set_attributes_in_dir<P>(
    dir_path: P,
    set_owner: &Option<String>,
    set_group: &Option<String>,
) -> Result<(), &'static str>
where
    P: AsRef<Path>,
{
    if set_owner.is_none() && set_group.is_none() {
        return Ok(());
    }
    let paths =
        read_dir(dir_path.as_ref()).map_err(|_| "Failed to list directory via indicated path.")?;
    for p in paths {
        if let Ok(ref entry) = p {
            if let Ok(ref ftype) = entry.file_type() {
                if ftype.is_file() {
                    let fpath = entry.path();
                    if let Some(ref new_owner) = set_owner {
                        file_owner::set_owner(&fpath, new_owner.as_str())
                            .map_err(|_| "Error setting new file owner")?;
                    }
                    if let Some(ref new_group) = set_group {
                        file_owner::set_group(&fpath, new_group.as_str())
                            .map_err(|_| "Error setting new file group")?;
                    }
                }
            }
        }
    }
    Ok(())
}

pub fn contains_unfinished_downloads<P>(dir_path: P) -> Result<bool, &'static str>
where
    P: AsRef<Path>,
    PathBuf: From<P>,
{
    let mut tmp_path: PathBuf = PathBuf::from(dir_path);
    tmp_path.push(TEMP_DIR);
    let paths = read_dir(tmp_path).map_err(|_| "Failed to list directory via indicated path.")?;
    let tmp_count = paths.into_iter().count();
    Ok(tmp_count > 0)
}

// TODO: implement
pub fn get_unfinished_downloads<P>(_dir_path: P) -> Result<HashSet<VideoId>, &'static str>
where
    P: AsRef<Path>,
    PathBuf: From<P>,
{
    todo!()
}

pub fn remove_tmp_if_empty<P>(base_path: P) -> Result<(), &'static str>
where
    P: AsRef<Path>,
    PathBuf: From<P>,
{
    let mut tmp_path: PathBuf = PathBuf::from(base_path);
    tmp_path.push(TEMP_DIR);
    let mut paths =
        read_dir(&tmp_path).map_err(|_| "Failed to list directory via indicated path.")?;
    if paths.next().is_none() {
        remove_dir(&tmp_path).map_err(|_| "Failed to remove temp directory.")?;
    } else {
        warn!("Can't remove temp dir, {:?} is not empty", &tmp_path);
    }
    Ok(())
}

/// Process single provided task.
pub async fn process_task(
    task: &DownloadEntity,
    set_owner: &Option<String>,
    set_group: &Option<String>,
) -> Result<(), &'static str> {
    let (send, recv) = oneshot::channel::<()>();
    let mut child = Command::new("yt-dlp")
        .stderr(Stdio::piped())
        .stdout(Stdio::piped())
        .args(&[
            "--no-progress",
            "-P",
            &task.output_path,
            "-P",
            format!("temp:{}", TEMP_DIR).as_str(),
        ])
        .args(&task.get_extra_flags())
        .arg(&task.url)
        .spawn()
        .map_err(|_| "Failed to spawn yt-dlp subprocess.")?;
    let stdout = child
        .stdout
        .take()
        .ok_or("Failed to get stdout handle from yt-dlp subprocess.")?;
    let stderr = child
        .stderr
        .take()
        .ok_or("Failed to get stderr handle from yt-dlp subprocess.")?;
    let mut reader_stdout = BufReader::new(stdout).lines();
    let mut reader_stderr = BufReader::new(stderr).lines();

    tokio::select! {
        retcode = child.wait() => {
            trace!("yt-dlp subprocess exited with: {:?}", retcode);
        },
        _ = async move {
            let send_cell = Cell::new(Option::Some(send));
            while let Ok(Some(line)) = reader_stdout.next_line().await {
                trace!("{}", line);
                if line.trim_end().ends_with("has already been downloaded") {
                    debug!("Found video that has already been downloaded, stopping.");
                    let _ = match send_cell.take() {
                        Some(s) => s.send(()),
                        _ => Ok(()),
                    };
                }
            }
        } => {},
        _ = async move {
            while let Ok(Some(line)) = reader_stderr.next_line().await {
                warn!("{}", line);
            }
        } => {},
        _ = recv => child.kill()
                       .await
                       .map_err(|_| "Failed to stop yt-dlp subprocess.")?
    }
    let set_attr_res = set_attributes_in_dir(&task.output_path, set_owner, set_group);
    let remove_tmp_res = remove_tmp_if_empty(&task.output_path);
    set_attr_res?;
    remove_tmp_res?;

    Ok(())
}

/// Iterate over available tasks and process each one linearly.
pub async fn process_all(
    tasks: &[DownloadEntity],
    set_owner: &Option<String>,
    set_group: &Option<String>,
) {
    // TODO: possibility for concurrent downloads?
    let mut counter = 0u32;
    for task in tasks.iter() {
        if let Err(e) = process_task(task, set_owner, set_group).await {
            error!("Task reported error: {}", e);
        }
        counter += 1;
    }
    info!("All tasks done, processed {} tasks", counter);
}

pub struct Watcher {
    settings: Settings,
}

impl Watcher {
    pub fn new(settings: Settings) -> Self {
        Watcher { settings }
    }

    /// Executes indefinitely, processing tasks according to schedule.
    pub async fn run(&self) -> Result<(), &'static str> {
        let (send, mut recv) = mpsc::channel::<()>(1);
        let scheduler = JobScheduler::new()
            .await
            .map_err(|_| "Failed to create scheduler")?;

        if let Some(update_now) = self.settings.update_on_start {
            if update_now {
                let res = send.send(()).await;
                if res.is_err() {
                    error!("Can't send scheduler ping: {}", res.unwrap_err());
                }
            }
        }

        for sched in self.settings.update_schedule.iter() {
            let send = send.clone();
            let cron_str = match sched {
                ScheduledTime::Cron(c) => c.clone(),
                ScheduledTime::Daily(t) => {
                    // TODO: this code is in dire need of pre-processed settings
                    if let Some((h, m)) = parse_time(t) {
                        format!("0 {} {} * * *", m, h)
                    } else {
                        error!("Unable to parse time \"{}\", skipping", t);
                        continue;
                    }
                }
            };
            let job = Job::new_async(cron_str.as_str(), move |_uuid, _lock| {
                let send = send.clone();
                Box::pin(async move {
                    debug!("Scheduler ping");
                    let res = send.send(()).await;
                    if res.is_err() {
                        error!("Can't send scheduler ping: {}", res.unwrap_err());
                    }
                })
            })
            .map_err(|_| "Failed to create async job")?;
            scheduler
                .add(job)
                .await
                .map_err(|_| "Failed to add a job to scheduler")?;
        }

        scheduler
            .start()
            .await
            .map_err(|_| "Failed to start scheduler")?;
        loop {
            // block current thread until scheduler ping
            if (recv.recv().await).is_some() {
                process_all(
                    &self.settings.tasks,
                    &self.settings.set_owner,
                    &self.settings.set_group,
                )
                .await;
            }
        }
    }
}


src/settings.rs
use serde::{Deserialize, Serialize};
use std::{fs::File, io::BufReader};
use tracing::debug;

static DEFAULT_AUDIO_FORMAT: &str = "mp3";

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DownloadEntity {
    /// Url for video or channel, any resource which yt-dlp supports
    pub url: String,
    /// Path where downloaded videos will be stored
    pub output_path: String,
    /// Human readable tag for this resource
    pub name: String,
    /// Extract audio, if `audio_format` is not set then default is mp3
    pub audio_only: Option<bool>,
    /// Has effect when `audio_only=true`, define output audio file format,
    /// possible options: mp3, aac, m4a, opus, vorbis, flac, alac, wav
    pub audio_format: Option<String>,
    /// Limit max resolution (i.e. video height), has no effect when `audio_only=true`
    /// examples: 480, 720, 1080, 1440, 2160, 4320
    pub max_resolution: Option<u32>,
    /// Limit download speed, used to prevent hogging a network connection
    /// examples: 3.5M, 300K
    pub throttle_speed: Option<String>,
    /// Custom flags to pass for yt-dlp, correctness is not checked
    pub custom_flags: Option<Vec<String>>,
}

impl DownloadEntity {
    pub fn get_extra_flags(&self) -> Vec<String> {
        let mut flags = Vec::new();
        if let Some(is_audio_only) = self.audio_only {
            if is_audio_only {
                let format = self
                    .audio_format
                    .as_ref()
                    .map(String::as_str)
                    .unwrap_or(DEFAULT_AUDIO_FORMAT);
                flags.extend(
                    ["-f", "ba", "-x", "--audio-format", format]
                        .iter()
                        .map(|&v| v.to_string()),
                );
            }
        } else if let Some(max_height) = self.max_resolution {
            flags.extend(
                [
                    "-f".to_string(),
                    format!("bestvideo[height<={}]+bestaudio", max_height),
                ]
                .into_iter(),
            );
        }
        if let Some(speed_limit) = self.throttle_speed.clone() {
            flags.extend(["-r".to_string(), speed_limit].into_iter());
        }
        if let Some(custom) = self.custom_flags.clone() {
            flags.extend(custom.into_iter());
        }
        debug!("extra flags: {:?}", flags);
        flags
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Settings {
    /// Videos/Channels to download
    pub tasks: Vec<DownloadEntity>,
    /// Schedule for running tasks in cron format
    pub update_schedule: Vec<ScheduledTime>,
    /// Custom log path
    pub log_dir: Option<String>,
    /// Force run tasks on start of program, regardless of schedule
    pub update_on_start: Option<bool>,
    /// Log level: Error, Warn, Info, Debug, Trace
    pub log_level: Option<LogLevel>,
    /// Set file owner for downloaded files
    pub set_owner: Option<String>,
    /// Set file group for downloaded files
    pub set_group: Option<String>,
}

/// Parse settings from json file, panics on error
pub fn load_settings(path: &str) -> Settings {
    let settings_file = File::open(path)
        .map_err(|e| eprintln!("Failed to locate settings file: {}", e))
        .unwrap();
    let reader = BufReader::new(settings_file);
    let settings = serde_json::from_reader(reader)
        .map_err(|e| eprintln!("Failed to deserialize settings file contents: {}", e))
        .unwrap();
    settings
}

/// Perform preliminary checks to ensure that settings values are well-formed
pub fn validate_settings(settings: &Settings) -> Result<(), &'static str> {
    if settings.tasks.is_empty() {
        return Err("Tasks list is empty.");
    }
    Ok(())
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum ScheduledTime {
    /// Execute single time, at specified time in format `hour:minute`
    Daily(String),
    /// Custom execution rule in cron format
    Cron(String),
}

/// Parse string in format `hour:minute` and return (hour, minute),
/// return None if parsing has failed
pub fn parse_time(value: &str) -> Option<(u8, u8)> {
    let nums: Vec<_> = value.split(":").collect();
    if nums.len() != 2 {
        return None;
    }
    let hour = nums[0].parse().ok()?;
    let minute = nums[1].parse().ok()?;
    Some((hour, minute))
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum LogLevel {
    Off,
    Error,
    Warn,
    Info,
    Debug,
    Trace,
}

impl From<LogLevel> for tracing::metadata::LevelFilter {
    fn from(level: LogLevel) -> Self {
        use LogLevel::*;
        match level {
            Off => Self::OFF,
            Error => Self::ERROR,
            Warn => Self::WARN,
            Info => Self::INFO,
            Debug => Self::DEBUG,
            Trace => Self::TRACE,
        }
    }
}
